### PR TITLE
Feature/create page save function

### LINKED ISSUE #14: Create page / save function / part 1
> This user story was split because the PR became too big. Part 2 can be found [here](https://github.com/kai-franke/capstone-project/issues/22).

## Value Proposition
As a tutor, I want the tutorial data I enter to conform to the specified format so that it can be saved in the future.

## Description
capstone-userstory-08

## Acceptance Criteria
- [x] shows a `finish creating` button
- [x] when this button is clicked an object is created which holds the entered data 
- [x] the input field `Tutorial title` must not be empty, must contain at least 5 and at most 60 characters
- [x] the input field `Step title` must not contain more than 60 characters
- [x] the input field `Step description` must not contain more than 300 characters
- [x] the input field `Picture URL` may only accept input in the form of a valid URL, which must end on `.gif`, `.jpg`, `.jpeg`, `.jfif`, `.pjpeg`, `.pjp`, `.png`, `.webp`.

## Tasks
- [x] make `Tutorial title` required, set minlength and maxlength
- [x] set maxlength for `Step title` 
- [x] set maxlength for `Step description`
- [x] set `type='url` and set pattern for `Picture URL`
- [x] add a `submit` button and name it `finish creating` 
- [x] make use of the entered data and convert it to the following structure:
```
{
    id: nanoid(),
    name: "Tutorial name",
    cover: "url of last step's image",
    slug: "tutorial-name-converted-to-lowercase-and-with-dashes",
    steps: [
        {
            step: 0,
        },
        {
            step: 1,
            title: "Step 1 title",
            description: "Step 1 description text",
            img: "Step 1 image url",
        },
        ...
    ],
}
```

### PR DESCRIPTION
**Pull request by kai-franke**
This PR checks and validates the values of the inputs when "finish creating" is clicked. In addition, the data is put into the required form to be able to write it to the localStorage later.
I intentionally left a console.log in the code to make it easier for Quality Assurance to check the finished dataset. 

I am very happy about your feedback!

**Preview:** https://capstone-project-git-feature-create-page-save-e3924a-kai-franke.vercel.app/create

### CODE REVIEW

File: components/CreateForm.js [hunk 1]
```diff
@@ -1,57 +1,118 @@
...
   }
 
+  function handleSubmit(event) {
+    event.preventDefault();
+    const formData = new FormData(event.target);
+    const data = Object.fromEntries(formData);
+    const slugSuffix = customAlphabet(
+      "23456789abcdefghklmnpqrstuvwxyzABCDEFGHKLMNPQRSTUVWXYZ",
+      4
+    );
```
Comment (ThomasF85) [inline components/CreateForm.js]: All functions and constants that don't depend on state or props can be outside of components. If they are inside a component they will be reinitiated every time the component renders. (Only makes sense if you don't want them to be reinitiated of course)

File: components/CreateForm.js [hunk 2]
```diff
@@ -60,7 +121,30 @@ function CreateForm() {
...
-          Add step
+          <IconContext.Provider
+            value={{
+              color: "inherit",
+              size: "1.4em",
+              title: "checkmark icon",
+              style: { marginRight: "0.5em" },
+            }}
+          >
+            <TbPlus />
```
Comment (ThomasF85) [inline components/CreateForm.js]: For a single item you can leave the IconContext away and just give the item the props directly. This also makes it more visible that you are not using styled components here.

**Review by ThomasF85 — APPROVED** [review]
Very nice! I just have two optional to implement comments.

**Comment by kai-franke** [comment]
> Very nice! I just have two optional to implement comments.

Thanks for your valuable feedback, @ThomasF85 !

File: components/CreateForm.js [hunk 3]
```diff
@@ -1,57 +1,142 @@
...
+const slugSuffix = customAlphabet(
+  "23456789abcdefghklmnpqrstuvwxyzABCDEFGHKLMNPQRSTUVWXYZ",
+  4
+);
 
 function CreateForm() {
-  const [steps, setSteps] = useState(1);
   const buttonRef = useRef();
+  const [inputSteps, setInputSteps] = useState([
+    { step: 1, stepTitle: "", stepUrl: "", stepDescription: "" },
```
Comment (F-Kirchhoff) [inline components/CreateForm.js]: maybe you can rename the stepUrl to stepImageUrl, that would be more descriptive

File: components/CreateForm.js [hunk 4]
```diff
@@ -1,57 +1,142 @@
...
-    });
+  function handleTitleChange(event) {
+    let titleInput = event.target.value;
+    // Thank you, https://github.com/Roland-Hufnagel
+    titleInput = titleInput.startsWith(" ")
+      ? titleInput.replace(" ", "")
+      : titleInput.includes("  ")
+      ? titleInput.replace("  ", " ")
+      : titleInput;
+    setInputTutorialTitle(titleInput);
```
Comment (F-Kirchhoff) [inline components/CreateForm.js]: Please use const here and define a new constant for the sanitized input named sanitizedInput for example.
You can also get rid of the ternary operators by just applying the sanitation without checking if they are necessary (if they are not the sanitation will yield the same title)
```
const sanitizedTitle = titleInput.trim().replace("  "," ")
```

File: components/CreateForm.js [hunk 5]
```diff
@@ -1,57 +1,142 @@
...
+
+  function handleFormChange(index, event) {
+    const data = [...inputSteps];
+    if (event.target.value.trim() === "") {
+      data[index][event.target.name] = "";
+    } else if (event.target.value[0] === " ") {
+      data[index][event.target.name] = event.target.value.trim();
+    } else {
+      data[index][event.target.name] = event.target.value;
+    }
```
Comment (F-Kirchhoff) [inline components/CreateForm.js]: This does exactly the same thing as your ternary operators above. Suggestion: create function called sanitizeString which does the sanitation as i described above and use it here as well!

File: components/CreateForm.js [hunk 6]
```diff
@@ -1,57 +1,142 @@
...
 
   function handleAddStep() {
-    setSteps((prevSteps) => prevSteps + 1);
+    const additionalStep = {
+      step: inputSteps.length + 1,
+      stepTitle: "",
+      stepUrl: "",
+      stepDescription: "",
+    };
+    setInputSteps((prevInputSteps) => [...prevInputSteps, additionalStep]);
```
Comment (F-Kirchhoff) [inline components/CreateForm.js:45]: nice!

File: components/CreateForm.js [hunk 7]
```diff
@@ -1,57 +1,142 @@
...
     scrollToButton();
   }
 
+  function handleSubmit(event) {
+    event.preventDefault();
+    if (inputTutorialTitle.replace(/[^a-zA-Z0-9]/g, "").length < 5) {
```
Comment (F-Kirchhoff) [inline components/CreateForm.js:51]: wow not bad

File: components/CreateForm.js [hunk 8]
```diff
@@ -1,57 +1,142 @@
...
+      return;
+    }
+    const newTutorial = {
+      id: nanoid(),
+      name: inputTutorialTitle,
+      cover: inputSteps[inputSteps.length - 1]["stepUrl"],
+      slug: inputTutorialTitle
+        .toLowerCase()
+        .replace(/ /g, "-")
+        .replace(/[-]+/g, "-")
```
Comment (F-Kirchhoff) [inline components/CreateForm.js]: I guess you can combine both regex as follows: `replace(/[ ]+/g, "-")`. But I am not 100% sure.

File: components/CreateForm.js [hunk 9]
```diff
@@ -1,57 +1,142 @@
...
+        .replace(/[^\w-]+/g, "")
+        .concat("-", slugSuffix()),
+      steps: [...inputSteps],
```
Comment (F-Kirchhoff) [inline components/CreateForm.js]: you don't need to make a copy here from the inputsteps:
```suggestion
      steps: inputSteps,
```

File: components/CreateForm.js [hunk 10]
```diff
@@ -1,57 +1,142 @@
...
+    };
+  }
+
+  function scrollToButton() {
+    setTimeout(() => {
+      buttonRef.current.scrollIntoView({ behavior: "smooth" });
+    });
+  }
+
```
Comment (F-Kirchhoff) [inline components/CreateForm.js:75]: nice idea!

File: components/CreateForm.js [hunk 11]
```diff
@@ -1,57 +1,142 @@
...
               </StyledLabel>
               <StyledLabel isPrimary={false}>
                 <LabelText>Step description</LabelText>
                 <StyledTextarea
+                  name="stepDescription"
+                  value={step.stepDescription}
                   placeholder="Enter a description"
                   aria-placeholder="Enter a description"
+                  maxLength="300"
+                  onChange={(event) => handleFormChange(index, event)}
```
Comment (F-Kirchhoff) [inline components/CreateForm.js:133]: I really like how you generalized one function for 3 inputs!

File: components/Footer.js [hunk 12]
```diff
@@ -25,6 +25,21 @@ function Footer() {
             Library
           </NavItem>
         </Link>
+        <Link href="/create" passHref>
+          <NavItem isActive={pathname === "/create"}>
```
Comment (F-Kirchhoff) [inline components/Footer.js:29]: nice!

**Review by F-Kirchhoff — CHANGES_REQUESTED** [review]
Very good code! I only found one spot where you could improve this PR. Have a look!

File: components/CreateForm.js (hunk 4 above)
Comment (kai-franke) [reply components/CreateForm.js]: That's awesome, @F-Kirchhoff ! trim() wouldn't let me use whitespaces anymore but trimStart() seems to do the job. :)

``` 
const sanitizedTitleInput = titleInput.trimStart().replace("  ", " ");
```

File: components/CreateForm.js (hunk 8 above)
Comment (kai-franke) [reply components/CreateForm.js]: Looks like it works. Thanks! 😄

**Review by ThomasF85 — APPROVED** [review]
Looks good!

**Pull request merged**